### ISSUE TITLE
Implementing a custom Field

### ISSUE THREAD

**Issue by feralvam** [allenai/allennlp#2052]
**System:**
 - OS: OSX
 - Python version: 3.6.6
 - AllenNLP version: I installed from master
 - PyTorch version: 0.4.1

**Question**
I am re-implementing part of the Entity Linking model described [here](https://arxiv.org/abs/1808.07699) ([TF code](https://github.com/dalab/end2end_neural_el)) using the AllenNLP framework. I am fairly new with using AllenNLP, so I would appreciate some recommendations on how to go about implementing one particular feature of the model. 

The model uses a file (_p_e_m.txt_) with lines  of the following structure:

```
<mention>\t<entity_id_1>,<entity_score_1>,<entity_name_1>\t<entity_id_2>,<entity_score_2>,<entity_name_2>\t...\t<entity_id_n>,<entity_score_n>,<entity_name_n>
```

Basically, for a mention, you have a list of candidate entities, each of them having an id, a probability and the entity name.

Now, for the model itself, you are given a document as input from which you extract spans (mentions) and you search in _p_e_m.txt_ for the corresponding candidate entities. You collect up to M candidate entities for each span you extract from the input document (not all spans are going to necessarily have M candidate entities). With this information, the model is then going to use pre-trained entity embeddings to compute some similarity and attention scores to predict which is the right entity for each extracted span (mention).

My question is: **what would be the best approach for loading these candidate entities for each mention (span) extracted from each document (instance)?**

For now, I am doing the following:

- In the `DatasetReader`, when I extract the spans of a document (in `text_to_instance`), I search in _p_e_m.txt_ for the candidate entities of each span. Now, I need a `Field` to store that list of candidate entities for each span. I am using a `ListField` for each span, which contains several `Field`s of type `EntityField` (a custom `Field` I implemented). Then, the whole instance contains a `ListField` of those `ListField`s (i.e., each span has a list of candidate entities, and the whole instance contains a list of those lists).

-  An `EntityField` contains only two attributes: the entity_name and the entity_score. The entity_name is a `Token`, while the entity_score is just a float. Why is the entity_name a `Token`? My reasoning is that since I need to load pre-trained embeddings for each entity, handling it in this manner I can take advantage of what AllenNLP already has implemented for word embeddings (each entity is just "a word"): in the `forward` method of the model I can just use a `TextFieldEmbedder` to get the embeddings for each candidate entity.

- What I am having trouble understanding now is how to implement the methods that a `Field` should override. Since in the`forward` method of my model I need both the entity embeddings and the scores, `batch_tensors` and `as_tensor` return dictionaries (I was inspired by `KnowledgeGraphField`): one item in the dict is the entity_name and the other is the score. However, I don't know how to handle `index` and  `get_padding_lengths`. I presume I need to re-implement `index` since I do have strings (entity_name) that need to be converted into indices, but I am not sure if I need to do something with `get_padding_lengths` and how this will affect `as_tensor`.

As you can see, I am a bit confused, and I would appreciate all the help you could provide. Also, if this is not the place to ask for this kind of question, let me know. Thank you in advance.

**Comment by matt-gardner** [comment]
This is the right place to ask this question, and thanks for giving sufficient detail that we can answer it =).

A way that might be slightly easier for you, without needing any custom fields, is to just have two parallel data structures, one containing the entity names, the other containing the entity probabilities.  That is, have a `ListField[ListField[TextField]]` for the entity names, and a `ListField[ListField[ArrayField]]` for the entity scores (you could also just do a straight `ArrayField` for the scores; it should come out the same.  But this way we do the padding for you.).  If you do this, you don't have to implement anything yourself except the dataset reader.

But, if you want to keep your custom `EntityField`, I'd basically just copy the logic that we have for the `TextField` and the `ArrayField` for each attribute (there's logic for both `index` and `get_padding_lengths` in `TextField`; `ArrayField` just has `get_padding_lengths`).  Because those are your data types, it should just work.

**Comment by joelgrus** [comment]
off the top of my head, I probably wouldn't create a custom field, which seems to be making your life more complicated; why not just do something like

```
spans: ListField[SpanField]   # one per span
entity_names: ListField[ListField[TextField]]  # one per entity per span
entity_scores: ListField[ArrayField]    # one array (with one entry per entity) per span                 
```

and then you don't have to worry about how to implement your own fields?

**Comment by joelgrus** [comment]
oops, jinx

**Comment by feralvam** [comment]
Thank you both for the quick replies. I have modified my implementation following your suggestions and it works well. 

The only problem I found was that for some mentions (spans) there are no candidate entities, but creating a `ListField` with an empty list produces an error. What I did was to create a dummy object of a `TextField` and a `ArrayField` and call their respective `empty_field` methods to create one element lists. 

Once again, thanks for the replies and for the great library in general.
